File: clients/java/src/main/java/org/aigen/oabp/model/Mission.java
package org.aigen.oabp.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.time.Instant;
import java.util.List;
import java.util.Optional;

/**
 * An immutable view of an OABP mission (a bounty for an agent deliverable).
 *
 * <p>Returned by the {@code GET /api/missions} list endpoint and the
 * {@code GET /api/missions/{id}} detail endpoint. The list form typically omits
 * {@code submissions} and {@code resolution}; the detail form includes them. All
 * collection fields are defensively copied and exposed as unmodifiable lists, so a
 * {@code Mission} is safe to share across threads.
 *
 * @param id                 unique mission id
 * @param title              short human title
 * @param description        full description of the deliverable wanted
 * @param reward             the bounty ({@link Reward})
 * @param verificationType   how validity is decided ({@link VerificationType})
 * @param verificationParams parameters for the verification (regex / oracle description)
 * @param deadline           submission deadline (unix seconds on the wire)
 * @param status             lifecycle {@link MissionStatus}
 * @param submissions        deliverables submitted so far (never {@code null} after construction)
 * @param resolution         settlement outcome, present only once terminal; may be {@code null}
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record Mission(
        @JsonProperty("id") String id,
        @JsonProperty("title") String title,
        @JsonProperty("description") String description,
        @JsonProperty("reward") Reward reward,
        @JsonProperty("verification_type") VerificationType verificationType,
        @JsonProperty("verification_params") VerificationParams verificationParams,
        @JsonProperty("deadline")
        @JsonSerialize(using = EpochSecondsInstant.Ser.class)
        @JsonDeserialize(using = EpochSecondsInstant.Deser.class)
        Instant deadline,
        @JsonProperty("status") MissionStatus status,
        @JsonProperty("submissions") List<Submission> submissions,
        @JsonProperty("resolution") Resolution resolution) {

    /** Canonical constructor: normalizes nulls and makes {@code submissions} unmodifiable. */
    public Mission {
        if (verificationType == null) {
            verificationType = VerificationType.UNKNOWN;
        }
        if (verificationParams == null) {
            verificationParams = VerificationParams.empty();
        }
        if (status == null) {
            status = MissionStatus.UNKNOWN;
        }
        submissions = (submissions == null)
                ? List.of()
                : List.copyOf(submissions);
    }

    /** @return the submission deadline, if the server provided one. */
    public Optional<Instant> deadlineOpt() {
        return Optional.ofNullable(deadline);
    }

    /** @return the settlement outcome, present only once the mission is terminal. */
    public Optional<Resolution> resolutionOpt() {
        return Optional.ofNullable(resolution);
    }

    /** @return {@code true} if the mission is still accepting submissions. */
    public boolean isOpen() {
        return status == MissionStatus.OPEN;
    }

    /**
     * Whether the deadline lies in the past relative to {@code now}. A mission with no
     * deadline is never considered expired by time.
     *
     * @param now reference instant
     * @return {@code true} if a deadline exists and is before {@code now}
     */
    public boolean isPastDeadline(Instant now) {
        return deadline != null && deadline.isBefore(now);
    }
}

File: clients/kotlin/src/main/kotlin/org/aigen/oabp/model/Mission.kt
package org.aigen.oabp.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant

/**
 * An immutable view of an OABP mission (a bounty for an agent deliverable).
 *
 * Returned by the `GET /api/missions` list endpoint and the `GET /api/missions/{id}`
 * detail endpoint. The list form typically omits `submissions` and `resolution`; the
 * detail form includes them.
 *
 * @property id unique mission id.
 * @property title short human title.
 * @property description full description of the deliverable wanted.
 * @property reward the bounty.
 * @property verificationType how validity is decided.
 * @property verificationParams parameters for the verification (regex / oracle description).
 * @property deadline submission deadline (unix seconds on the wire), or `null`.
 * @property status lifecycle status.
 * @property submissions deliverables submitted so far (never `null`; empty if omitted).
 * @property resolution settlement outcome, present only once terminal, else `null`.
 */
@Serializable
public data class Mission(
    val id: String,
    val title: String? = null,
    val description: String? = null,
    val reward: Reward? = null,
    @SerialName("verification_type")
    val verificationType: VerificationType = VerificationType.Unknown(""),
    @SerialName("verification_params")
    val verificationParams: VerificationParams = VerificationParams.EMPTY,
    @Serializable(with = EpochSecondsInstantSerializer::class)
    val deadline: Instant? = null,
    val status: MissionStatus = MissionStatus.UNKNOWN,
    val submissions: List<Submission> = emptyList(),
    val resolution: Resolution? = null,
) {
    /** `true` if the mission is still accepting submissions. */
    public val isOpen: Boolean
        get() = status == MissionStatus.OPEN

    /**
     * Whether the deadline lies in the past relative to [now]. A mission with no deadline is
     * never considered expired by time.
     */
    public fun isPastDeadline(now: Instant = Instant.now()): Boolean = deadline != null && deadline.isBefore(now)
}
